Source/Core/VideoCommon/PerformanceTracker.h

```diff
@@ -0,0 +1,105 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+#include <array>
+#include <chrono>
+#include <fstream>
+#include <mutex>
+#include <optional>
+
+#include "Common/CommonTypes.h"
+
+class PerformanceTracker
+{
+private:
+  // Must be powers of 2 for masking to work
+  static constexpr u64 MAX_DT_QUEUE_SIZE = 1UL << 12;
+  static constexpr u64 MAX_QUALITY_GRAPH_SIZE = 1UL << 8;
+
+  static inline std::size_t IncrementIndex(const std::size_t index)
+  {
+    return (index + 1) & (MAX_DT_QUEUE_SIZE - 1);
+  }
+
+  static inline std::size_t DecrementIndex(const std::size_t index)
+  {
+    return (index - 1) & (MAX_DT_QUEUE_SIZE - 1);
+  }
+
+  static inline std::size_t GetDifference(const std::size_t begin, const std::size_t end)
+  {
+    return (end - begin) & (MAX_DT_QUEUE_SIZE - 1);
+  }
+
+public:
+  PerformanceTracker(const char* log_name = nullptr,
+                     const std::optional<s64> sample_window_us = {});
+  ~PerformanceTracker();
+
+  PerformanceTracker(const PerformanceTracker&) = delete;
+  PerformanceTracker& operator=(const PerformanceTracker&) = delete;
+  PerformanceTracker(PerformanceTracker&&) = delete;
+  PerformanceTracker& operator=(PerformanceTracker&&) = delete;
+
+public:  // Functions for recording and accessing information
+  void Reset();
+  void Count();
+
+  DT GetSampleWindow() const;
+
+  double GetHzAvg() const;
+
+  DT GetDtAvg() const;
+  DT GetDtStd() const;
+
+  DT GetLastRawDt() const;
+
+  void ImPlotPlotLines(const char* label) const;
+
+private:  // Functions for managing dt queue
+  inline void QueueClear();
+  inline void QueuePush(DT dt);
+  inline const DT& QueuePop();
+  inline const DT& QueueTop() const;
+  inline const DT& QueueBottom() const;
+
+  std::size_t inline QueueSize() const;
+  bool inline QueueEmpty() const;
+
+private:  // Handle pausing and logging
+  void LogRenderTimeToFile(DT val);
+  void SetPaused(bool paused);
+
+private:
+  // Handle not counting time during pauses
+  bool m_paused = false;
+  int m_on_state_changed_handle;
+
+  // Name of log file and file stream
+  const char* m_log_name;
```

Comment:
IMO having this be std::string (and thus having ownership of the string data) would be better, vs. the user of this object having to maintain the filename alive until this object is deleted. Right now it doesn't make a huge difference because (I think) this is only ever being used with constant strings, but in the future it could lead to confusion.